### PR TITLE
Avoid using a webpack builded file for static build

### LINKED ISSUE #23: Dev server should generate static version for url on the fly
Currently we need to build a static version + run the dev server. This is counter productive. The express server should be able to generate the page on the fly directly.

### LINKED ISSUE #92: Bundle collection in window.__INITIAL_STATE__
I followed your instruction on create a collection page [here](http://moox.io/statinamic/docs/usage/collections/)

[Here is my collection source code](https://github.com/thangngoc89/blog/blob/master/web_modules%2Fcontainers%2FArchive%2FArchive.js)

But React will throw a warning:

```
Warning: React attempted to reuse markup in a container but the checksum was invalid. This generally means that you are using server rendering and the markup generated on the server was not what the client was expecting. React injected new markup to compensate which works but you have lost many of the benefits of server rendering. Instead, figure out why the markup being generated is different on the client or server:
 (client) MgeFRiHrav1TDm192y99 _3jzDdN3QzHnR0O_efk
 (server) MgeFRiHrav1TDm192y99" href="/" data-reac
```

Demo: http://khoanguyen.me (open your console)

After poking around your code, I see the issue is from this line 
https://github.com/MoOx/statinamic/blob/master/src%2Fstatic%2Fto-html%2Furl-as-html.js#L75

Is it posible that we could pass an option to bundle `collection` in `INITIAL_STATE` on specific page ?

### PR DESCRIPTION
**Pull request by MoOx**
By using webpack on the fly, we can now pre-render page during
development. And static build is faster.

Will close #23, close #92, close #100, close #106 

I stil have 2 things to fix (some tests broken + css file link in static html)

@thangngoc89 you can already take a look and try if you want :)

### CODE REVIEW

**Comment by thangngoc89** [comment]
:+1

Omg.  It's huge

**Comment by thangngoc89** [comment]
@MoOx there is some bug that prevent react bootstrap on client side:

```
Warning: React.createElement: type should not be null, undefined, boolean, or number. It should be a string (for DOM elements) or a ReactClass (for composite components).warning @ (program):45
0.0.0.0/:46 Uncaught Invariant Violation: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined.
```

Repro:

- Fresh clone
- npm install
- cd docs && npm install
- npm i -D raw-loader react-svg-inline
- npm install ../
- npm start

File: src/client/index.js [hunk 1]
```diff
@@ -2,43 +2,34 @@
...
-        error,
-      })
-    )
-
   ReactDOM.render(
     <div id="statinamic-container">
-      <MetadataProvider metadata={ metadata }>
-        <Provider store={ store }>
+      <StatinamicContextProvider
+        collection={ typeof window !== undefined && window.__COLLECTION__ }
```
Comment (thangngoc89) [inline src/client/index.js]: `typeof window !== 'undefined'` ?

File: src/client/index.js [hunk 2]
```diff
@@ -2,43 +2,34 @@
 import React from "react"
 import ReactDOM from "react-dom"
 import Router, { browserHistory } from "react-router"
```
Comment (thangngoc89) [inline src/client/index.js]: This should be `{ Router, browserHistory } from ...`

File: src/client/index.js [hunk 3]
```diff
@@ -2,43 +2,34 @@
...
 
 export default function statinamic({
+  layouts,
   metadata,
   routes,
   store,
 }) {
   let devtools = false
-  if (__DEVTOOLS__) {
+  if (process.env.REDUX_DEVTOOLS) {
```
Comment (thangngoc89) [inline src/client/index.js]: There should be a check for window object since devtools require window

Comment (MoOx) [inline src/client/index.js]: The idea is to avoid this code when builded in production. A test on window. will make this build to have redux devtools code all the time and we don't want that. That's why I used process.env.* which can be replaced by webpack with DefinePlugin. That's said for now I only inject NODE_ENV, so I will add this one and document this.

**Comment by MoOx** [comment]
I have probably forget some deps since static build "works for me" :)
Will double check and update PR.

**Comment by thangngoc89** [comment]
`npm prune` to remove unlisted deps

File: src/client/index.js (hunk 3 above)
Comment (thangngoc89) [inline src/client/index.js]: This line https://github.com/MoOx/statinamic/blob/babel-node-loaders/src/redux/createStore/index.js#L33 should be 

```
process.env.REDUX_DEVTOOLS && 
typeof window !== "undefined"
```

because on line 38 of the same file. We are using window object which is undefined in static build.

Anyway, just run `REDUX_DEVTOOLS=true npm start` and you will see a bunch of error

Comment (MoOx) [inline src/client/index.js]: The thing is: we need a test that can be statically converted when uglifying. I will add a process.env.CLIENT then...

Comment (thangngoc89) [inline src/client/index.js]: :+1:

**Comment by MoOx** [comment]
npm pruned, branch rebased, commit updated/amended.

Use "npm test" instead of your combo (npm test use a symlink trick to make the things works properly, like in "real word", when statinamic is at the same level as other deps).

**Comment by thangngoc89** [comment]
You forgot to remove emitting collection.json in md-collection-loader

**Comment by MoOx** [comment]
Good catch.
I will remove it.
I think I will also add some "funtional" testing using the docs/dist folder. It's not easy to handle a full coverage for this kind of project :)
I have to go for today and will handle that this evening or tomorrow.

**Comment by thangngoc89** [comment]
Thank you for all of your hard works :+1: . Can't wait until this PR get merged

**Comment by thangngoc89** [comment]
No HMR for React components ?

**Comment by MoOx** [comment]
It seems that I have introduced a bug. I will review that.

**Comment by thangngoc89** [comment]
You removed babel-preset-react-hmre from source code

**Comment by MoOx** [comment]
I will add it back in boilerplate. That's because I add to remove it from global config.

**Comment by MoOx** [comment]
I just removed collection.json and fixed hmre.

**Comment by MoOx** [comment]
For npm I don't have anything magic. npm@3 is slow. Did you try latest 3.7.1 which is supposed to be faster?

**Comment by MoOx** [comment]
The fuck babel. This is very annoying. The issue "Cannot read property 'start' of undefined" is still here on another part of the code...

**Comment by thangngoc89** [comment]
Lock it to 6.4.5 ?

**Comment by MoOx** [comment]
Not a solution. Ok this bug just have been fixed in babel-plugin-webpack-loaders.

**Comment by thangngoc89** [comment]
Ha. That's why it's faliing on others project of mine

**Comment by MoOx** [comment]
Just use `"babel-plugin-webpack-loaders": "^0.2.2"`.

**Comment by thangngoc89** [comment]
This may be not related to this PR, but can you add some sourcemap ? It makes debugging much better

**Comment by MoOx** [comment]
No idea how to add that :)

**Comment by thangngoc89** [comment]
https://webpack.github.io/docs/configuration.html#devtool

tl;dr

Add this to webpack config on development:

`devtool: "cheap-module-eval-source-map"`

**Comment by MoOx** [comment]
Feel free to make a PR for devtools.
